File: AlgorithmJavaVersion/src/Algorithm/math/Permutation.java
package Algorithm.math;

import Top100.Mathlj;
import Common.Constant.C;

/**
 * @author 蔚蔚樱
 * @version 1.0
 * @date 2022/5/21 14:40
 * @author—Email micromicrohard@outlook.com
 * @description 数学 排列数
 * @blogURL
 */

public class Permutation implements Mathlj {

    public static int CalMethod(int m, int n) {
        if (n < 0 || m < n) {
            return C.ErrorNum;
        }
        return exec2(m, n);
    }

    public static int exec(int m, int n) {
        int m_Factorial = Factorial.CalMethod(m);
        int m_n_Factorial = Factorial.CalMethod(m - n);
        return m_Factorial / m_n_Factorial;
    }

    public static int exec2(int m, int n) {
        int result = 1;
        for (int i = m; i > m - n; i--) {
            result *= i;
        }
        return result;
    }
}

File: AlgorithmGoVersion/src/Algorithm/math/Permutation.go
package math

import C "AlgorithmPractice/src/Common/Constant"

// Permutation
// @author: 蔚蔚樱
// @date: 2022/5/2219:50
// @author—Email: micromicrohard@outlook.com
// @description: 数学 排列数
// @blogURL:
type Permutation struct {
}

// PermutationMethod = C(m,n)
func (p *Permutation) PermutationMethod(m, n int) int {
	if m < n || n < 0 {
		return C.ErrorNum
	}
	return p.exec2(m, n)
}

func (p *Permutation) exec(m, n int) int {
	f := &Factorial{}
	m_Factorial := f.FactorialMethod(m)
	n_Factorial := f.FactorialMethod(m - n)
	return m_Factorial / n_Factorial
}

func (p *Permutation) exec2(m, n int) int {

	result := 1
	for i := m; i > m-n; i-- {
		result *= i
	}
	return result
}
